torch/csrc/jit/tensorexpr/bounds_overlap.cpp

```diff
@@ -0,0 +1,256 @@
+#include <torch/csrc/jit/tensorexpr/bounds_overlap.h>
+#include <torch/csrc/jit/tensorexpr/ir_simplifier.h>
+
+namespace torch {
+namespace jit {
+namespace tensorexpr {
+namespace bounds {
+
+OverlapKind boundOverlap(Bound a, Bound b) {
+  // If they're equal they're equal.
+  bool startEqual = exprEquals(a.start, b.start);
+  bool endEqual = exprEquals(a.end, b.end);
+  if (startEqual && endEqual) {
+    return TotalOverlapB;
+  }
+
+  const Expr* aSize = IRSimplifier::simplify(new Sub(a.end, a.start));
+  const Expr* bSize = IRSimplifier::simplify(new Sub(b.end, b.start));
+
+  const Expr* lowDiff = IRSimplifier::simplify(new Sub(a.start, b.end));
+  const Expr* highDiff = IRSimplifier::simplify(new Sub(b.start, a.end));
+
+  if (lowDiff->isConstant() && highDiff->isConstant()) {
+    int low = immediateAs<int>(lowDiff);
+    int high = immediateAs<int>(highDiff);
+    // No overlap.
+    if (low > 0 || high > 0) {
+      return NoOverlap;
+    }
+  }
+
+  if ((highDiff->isConstant() && immediateAs<int>(highDiff) == 0) ||
+      (lowDiff->isConstant() && immediateAs<int>(lowDiff) == 0)) {
+    if (aSize->isConstant() && immediateAs<int>(aSize) <= 0) {
+      return TotalOverlapB;
+    }
```

Comment:
I'm not sure about this. Consider the following scenario where A1 is A-start, A2 is A-end, B1 is B-start, B2 is B-end.

-----A1--------------A2-------
-----B1-----------B2----------

This is "overlapping". However, as per your calculations here:
  `lowDiff = A1 - B2`
  `highDiff = B1 - A2`

Neither of those is going to be 0. Am I missing something here?

AFAICT, the differences here represents the 'adjacent' case and one of the diffs below, `diffs` or `diffe`, being 0 represents the 'overlapping' case.

Reply:
Let me add a test for this situation and verify.

Reply:
Yeah, I have the same question. This doesn't look conservative.